### ISSUE TITLE
Wrong length of strings returned by string_append_vformat

### ISSUE THREAD

**Issue by P-p-H-d** [ludocode/pottery#3]
For the following code, I was expecting the length computed by string\_length and by strlen to be the same:

```
#include <stdio.h>
#include "util/pottery/string/string.h"
int main(void)
{
  string_t s;
  string_init(&s);
  string_append_format(&s,"%u",245234345);
  printf("S=%s L0=%zu L1=%zu\n",
         string_cstr(&s),
         string_length(&s),
         strlen(string_cstr(&s)));
  string_clear(&s);
  return 0;
}
```

whereas it outputs:
`
S=245234345 L0=10 L1=9
`
i.e. the output of string\_length is one too many.

**Comment by P-p-H-d** [comment]
I think the issue is in string\_append\_vformat when it needs to redo the call to vsnprintf.
I propose the following patch:
```
--- a/util/pottery/string/string.c
+++ b/util/pottery/string/string.c
@@ -239,7 +239,7 @@ void string_append_vformat(string_t* string,
 
     // Otherwise grow to the necessary size and format again.
     buffer_length = pottery_cast(size_t, ret) + 1;
-    total_length = original_length + buffer_length;
+    total_length = original_length + buffer_length - 1;
     if (total_length < original_length)
         abort();
     string_set_length(string, total_length);
```

**Comment by ludocode** [comment]
Thanks for the bug report. I pushed a slightly different fix from what you did because I thought your code wasn't leaving room for the null-terminator. Of course the string has a null-terminator already in it.

I'm going to clean up this code a bit to make it faster. It should be using the storage directly instead of calling `string_set_length()` because it keeps having to unnecessarily re-adjust for the null-terminator (and in fact the code I just pushed ends up having two null-terminators at the end before removing one.) I also want to optimize `string_insert_vformat()`. Currently it always writes to an auxiliary buffer but this shouldn't be necessary because it's possible to make `vsnprintf()` omit the null-terminator just be giving it the exact length of the string without it.
